teos/src/main.rs

```diff
@@ -77,14 +81,66 @@ async fn main() {
     });
 
     // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
-    let mut conf = config::from_file::<Config>(path.join("teos.toml"));
+    let mut conf = config::from_file::<Config>(conf_file_path);
     let is_default = conf.is_default();
     conf.patch_with_options(opt);
     conf.verify().unwrap_or_else(|e| {
         eprintln!("{e}");
         std::process::exit(1);
     });
 
+    // log default data dir
+    log::info!("Using default data directory: {:?}", &path);
+
+    // log the non-default options that the daemon is using
+    if !is_default {
+        if conf.api_bind != "localhost" {
+            log::info!("Using non-default API bind address: {}", conf.api_bind);
+        }
+        if conf.api_port != 9814 {
+            log::info!("Using non-default API port: {}", conf.api_port);
+        }
+        if conf.rpc_bind != "localhost" {
+            log::info!("Using non-default RPC bind address: {}", conf.rpc_bind);
+        }
+        if conf.rpc_port != 8814 {
+            log::info!("Using non-default RPC port: {}", conf.rpc_port);
+        }
+        if conf.btc_network != "mainnet" {
+            log::info!("Using non-default Bitcoin network: {}", conf.btc_network);
+        }
+        if conf.btc_rpc_user != "user" {
+            log::info!("Using non-default Bitcoin RPC user: {}", conf.btc_rpc_user);
+        }
+        if conf.btc_rpc_password != "passwd" {
+            log::info!(
+                "Using non-default Bitcoin RPC password: {}",
+                conf.btc_rpc_password
+            );
+        }
+        if conf.btc_rpc_connect != "localhost" {
+            log::info!(
+                "Using non-default Bitcoin RPC connect address: {}",
+                conf.btc_rpc_connect
+            );
+        }
+        if conf.btc_rpc_port != 0 {
+            log::info!("Using non-default Bitcoin RPC port: {}", conf.btc_network);
+        }
+        if conf.tor_control_port != 9051 {
+            println!(
+                "Using non-default Tor control port: {}",
+                conf.tor_control_port
+            );
+        }
+        if conf.onion_hidden_service_port != 9814 {
+            println!(
+                "Using non-default onion hidden service port: {}",
+                conf.onion_hidden_service_port
+            );
+        }
+    }
+
```

Comment:
This is not the right way to handle this. We don't want to add all this boilerplate to main. Also, most of this info is already in Config. The right way to handle this will most likely involve the Default and Eq traits.

Reply:
> This is not the right way to handle this. We don't want to add all this boilerplate to main. Also, most of this info is already in Config. The right way to handle this will most likely involve the Default and Eq traits.

Please check the latest commit and let me know if it's correct or not.

Reply:
> Please check the latest commit and let me know if it's correct or not.

```rust
        if conf.btc_rpc_port != 0 {
            log::info!("Using non-default Bitcoin RPC port: {}", conf.btc_network);
        }
        if conf.btc_rpc_user != "user" {
            log::info!("Using non-default Bitcoin RPC user: {}", conf.btc_rpc_user);
        }
        if conf.btc_rpc_password != "passwd" {
            log::info!("Using non-default Bitcoin RPC password: {}", conf.btc_rpc_password);
        }
        if conf.btc_network != "mainnet" {
            log::info!("Using non-default Bitcoin network: {}", conf.btc_network);
        }
        if conf.btc_rpc_connect != "localhost" {
            log::info!("Using non-default Bitcoin RPC connect address: {}", conf.btc_rpc_connect);
        }
```

You are still using hard coded default values ("mainnet", "localhost", etc...) but you could create a default config struct and compare it against `conf` instead.

Reply:
> > Please check the latest commit and let me know if it's correct or not.
> 
> ```
>         if conf.btc_rpc_port != 0 {
>             log::info!("Using non-default Bitcoin RPC port: {}", conf.btc_network);
>         }
>         if conf.btc_rpc_user != "user" {
>             log::info!("Using non-default Bitcoin RPC user: {}", conf.btc_rpc_user);
>         }
>         if conf.btc_rpc_password != "passwd" {
>             log::info!("Using non-default Bitcoin RPC password: {}", conf.btc_rpc_password);
>         }
>         if conf.btc_network != "mainnet" {
>             log::info!("Using non-default Bitcoin network: {}", conf.btc_network);
>         }
>         if conf.btc_rpc_connect != "localhost" {
>             log::info!("Using non-default Bitcoin RPC connect address: {}", conf.btc_rpc_connect);
>         }
> ```
> 
> You are still using hard coded default values ("mainnet", "localhost", etc...) but you could create a default config struct and compare it against `conf` instead.

So, I should create a default create config struct in config.rs and match it main using conf?

Reply:
> > So, I should create a default config struct
> 
> Yeah. But it's already there. `Config` already implements `std::Default`

changed now. please let me know if it's ok now.

Reply:
You are still bloating main with the comparisons. Just implement a function in config that deals with the diff between the given `Config` and the default one. It could be something along the lines of a method that returns a `HashMap` where keys are the config names, and values are a tuple of the expected value and the received value. Then, you can simply iterate through the map in main and log accordingly.